test/index.ts

```diff
@@ -625,15 +625,29 @@ function testNoop() {
   const callback: (_: number) => void = noop;
 }
 
-function testHeadTail() {
+function testHead() {
   type List1 = [number, string, boolean, "a" | "b"];
 
   type TestHead = Assert<IsExact<Head<List1>, number>>;
   type TestHeadOnEmpty = Assert<IsExact<Head<[]>, never>>;
+}
 
-  type TestTail = Assert<IsExact<Tail<List1>, [string, boolean, "a" | "b"]>>;
-  type TestTailOnEmpty = Assert<IsExact<Tail<[]>, never>>;
-  type TestTailOn1Elem = Assert<IsExact<Tail<[number]>, []>>;
+function testTail() {
+  type cases = [
+    Assert<IsExact<Tail<[]>, never>>,
+    Assert<IsExact<Tail<readonly []>, never>>,
+    Assert<IsExact<Tail<[1, 2, 3 | 4]>, [2, 3 | 4]>>,
+    Assert<IsExact<Tail<readonly [1, 2, 3 | 4]>, [2, 3 | 4]>>,
+    Assert<IsExact<Tail<[1]>, []>>,
+    Assert<IsExact<Tail<[1?]>, []>>,
+    Assert<IsExact<Tail<[1, 2] | [3, 4]>, [2] | [4]>>,
+    Assert<IsExact<Tail<[1?, 2?, 3?]>, [2?, 3?]>>,
+    Assert<IsExact<Tail<[1, 2, ...number[]]>, [2, ...number[]]>>,
+    Assert<IsExact<Tail<number[]>, number[]>>,
+    Assert<IsExact<Tail<readonly number[]>, number[]>>,
+    Assert<IsExact<Tail<any>, unknown[]>>,
```

Comment:
(optional) this one is interesting, do you know why it equals to `unknown[]`? If you can document it above in the comment, it would be great

Reply:
@Beraliv If you think this makes sense, then I'll add a link to this next to the test case.

Reply:
Right, thank you for clarifying it for me!

> Now, Rest happens to be unknown[] probably because in the conditional Type extends readonly [any?, ...infer Rest], when Type is any, which is not an array, Rest falls back to its default constraint of unknown[]. A similar thing happens if Type is [].

This makes total sense, so it's a conditional type with inference that makes it `unknown[]`

> If we add an explicit constraint of any[] to Rest, then we'd get back any[] for any.

Please keep it as it is (because conditional constraint are not supported in all TypeScript versions that we support)

Reply:
> @Beraliv If you think this makes sense, then I'll add a link to this next to the test case.

Please do!

Reply:
Added it in [this commit](https://github.com/ts-essentials/ts-essentials/pull/424/commits/3ae6fdf5eaedaa32584e3f56446a1c5886791fb4). Also, added a test case for `any[]`.

Reply:
Also, included test cases for union of tuples with different lengths ([#e417a8c](https://github.com/ts-essentials/ts-essentials/pull/424/commits/e417a8c93ae2b675d5283b8e8b6df0ca98b92c9d)).

Because, this is the case (`[1, 2] | []`) where `Head` type fails, so it's worth having it for `Tail` as well.